src/ray/util/url.cc

```diff
@@ -16,64 +16,103 @@
 
 #include <stdlib.h>
 
+#ifdef _WIN32
+#include <afunix.h>
+#else
+#include <sys/un.h>
+#endif
+
 #include <algorithm>
-#include <boost/asio/ip/address.hpp>
+#include <boost/asio/generic/stream_protocol.hpp>
+#ifndef _WIN32
+#include <boost/asio/local/stream_protocol.hpp>
+#endif
 #include <boost/asio/ip/tcp.hpp>
+#include <sstream>
 #include <string>
 
+#include "ray/util/filesystem.h"
 #include "ray/util/logging.h"
+#include "ray/util/util.h"
 
-boost::asio::ip::tcp::endpoint parse_ip_tcp_endpoint(const std::string &endpoint,
-                                                     int default_port) {
-  const std::string scheme_sep = "://";
-  size_t scheme_begin = 0, scheme_end = endpoint.find(scheme_sep, scheme_begin);
-  size_t host_begin;
-  if (scheme_end < endpoint.size()) {
-    host_begin = scheme_end + scheme_sep.size();
-  } else {
-    scheme_end = scheme_begin;
-    host_begin = scheme_end;
-  }
-  std::string scheme = endpoint.substr(scheme_begin, scheme_end - scheme_begin);
-  RAY_CHECK(scheme_end == host_begin || scheme == "tcp");
-  size_t port_end = endpoint.find('/', host_begin);
-  if (port_end >= endpoint.size()) {
-    port_end = endpoint.size();
+namespace ray {
+
+std::string endpoint_to_url(
+    const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol> &ep,
+    bool include_scheme) {
+  std::string result;
+  switch (ep.protocol().family()) {
+  case AF_INET:
+  case AF_INET6: {
+    boost::asio::ip::tcp protocol = ep.protocol().family() == AF_INET
+                                        ? boost::asio::ip::tcp::v4()
+                                        : boost::asio::ip::tcp::v6();
+    boost::asio::ip::tcp::endpoint e(protocol, 0);
+    RAY_DCHECK(e.size() == ep.size());
+    memcpy(e.data(), ep.data(), ep.size());
+    std::ostringstream ss;
+    ss << e;
+    result = ss.str();
+    if (include_scheme) {
+      result.insert(0, "tcp://");
+    }
+    break;
   }
-  size_t host_end, port_begin;
-  if (endpoint.find('[', host_begin) == host_begin) {
-    // IPv6 with brackets (optional ports)
-    ++host_begin;
-    host_end = endpoint.find(']', host_begin);
-    if (host_end < port_end) {
-      port_begin = endpoint.find(':', host_end + 1);
-      if (port_begin < port_end) {
-        ++port_begin;
-      } else {
-        port_begin = port_end;
-      }
-    } else {
-      host_end = port_end;
-      port_begin = host_end;
+  case AF_UNIX: {
+    size_t path_offset = offsetof(sockaddr_un, sun_path);
```

Comment:
I feel like the above approach of `ep.address().to_string()` should also work for this code path, can you check? In that case they can be unified via
```
std::string address = ep.address().to_string();
if (ep.protocol().family() == AF_INET || ep.protocol().family() == AF_INET6) {
    return "tcp://" + address;
} else if (ep.protocol().family() == AF_UNIX) {
  return "unix:///" + address
} else {
  RAY_LOG(FATAL) << "unsupported protocol family: " << ep.protocol().family();
}
```

Reply:
[There's no `address()` method for generic sockets](https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/generic__basic_endpoint.html), only for specialized socket types (TCP, local), so it won't work here unfortunately.